


<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html id="htmlId">
<head>
  <title>Coverage Report :: BeautiAlignmentProvider</title>
  <style type="text/css">
    @import "../../.css/coverage.css";
  </style>
</head>

<body>
<div class="header"></div>

<div class="content">
<div class="breadCrumbs">
    [ <a href="../../index.html">all classes</a> ]
    [ <a href="../index.html">beast.app.beauti</a> ]
</div>

<h1>Coverage Summary for Class: BeautiAlignmentProvider (beast.app.beauti)</h1>

<table class="coverageStats">
<tr>
  <th class="name">Class</th>
<th class="coverageStat 
">
  Class, %
</th>
<th class="coverageStat 
">
  Method, %
</th>
<th class="coverageStat 
">
  Line, %
</th>
</tr>
<tr>
  <td class="name">BeautiAlignmentProvider</td>
<td class="coverageStat">
  <span class="percent">
    100%
  </span>
  <span class="absValue">
    (1/ 1)
  </span>
</td>
<td class="coverageStat">
  <span class="percent">
    13.3%
  </span>
  <span class="absValue">
    (2/ 15)
  </span>
</td>
<td class="coverageStat">
  <span class="percent">
    1.4%
  </span>
  <span class="absValue">
    (3/ 218)
  </span>
</td>
</tr>

</table>

<br/>
<br/>


<div class="sourceCode"><i>1</i>&nbsp;package beast.app.beauti;
<i>2</i>&nbsp;
<i>3</i>&nbsp;
<i>4</i>&nbsp;
<i>5</i>&nbsp;
<i>6</i>&nbsp;
<i>7</i>&nbsp;import java.io.BufferedReader;
<i>8</i>&nbsp;import java.io.File;
<i>9</i>&nbsp;import java.io.FileReader;
<i>10</i>&nbsp;import java.io.IOException;
<i>11</i>&nbsp;import java.io.StringReader;
<i>12</i>&nbsp;import java.util.ArrayList;
<i>13</i>&nbsp;import java.util.Collections;
<i>14</i>&nbsp;import java.util.HashMap;
<i>15</i>&nbsp;import java.util.HashSet;
<i>16</i>&nbsp;import java.util.List;
<i>17</i>&nbsp;import java.util.Map;
<i>18</i>&nbsp;import java.util.Set;
<i>19</i>&nbsp;
<i>20</i>&nbsp;import javax.swing.JOptionPane;
<i>21</i>&nbsp;import javax.xml.parsers.DocumentBuilderFactory;
<i>22</i>&nbsp;import javax.xml.parsers.ParserConfigurationException;
<i>23</i>&nbsp;
<i>24</i>&nbsp;import org.w3c.dom.Document;
<i>25</i>&nbsp;import org.w3c.dom.NodeList;
<i>26</i>&nbsp;import org.xml.sax.InputSource;
<i>27</i>&nbsp;import org.xml.sax.SAXException;
<i>28</i>&nbsp;
<i>29</i>&nbsp;import beast.core.BEASTInterface;
<i>30</i>&nbsp;import beast.core.BEASTObject;
<i>31</i>&nbsp;import beast.core.Description;
<i>32</i>&nbsp;import beast.core.Input;
<i>33</i>&nbsp;import beast.core.Input.Validate;
<i>34</i>&nbsp;import beast.evolution.alignment.Alignment;
<i>35</i>&nbsp;import beast.evolution.alignment.FilteredAlignment;
<i>36</i>&nbsp;import beast.evolution.alignment.Sequence;
<i>37</i>&nbsp;import beast.evolution.datatype.DataType;
<i>38</i>&nbsp;import beast.math.distributions.MRCAPrior;
<i>39</i>&nbsp;import beast.util.NexusParser;
<i>40</i>&nbsp;import beast.util.XMLParser;
<i>41</i>&nbsp;
<i>42</i>&nbsp;
<i>43</i>&nbsp;
<i>44</i>&nbsp;@Description(&quot;Class for creating new alignments to be edited by AlignmentListInputEditor&quot;)
<b class="fc"><i>45</i>&nbsp;public class BeautiAlignmentProvider extends BEASTObject {</b>
<i>46</i>&nbsp;	
<b class="fc"><i>47</i>&nbsp;	final public Input&lt;BeautiSubTemplate&gt; template = new Input&lt;&gt;(&quot;template&quot;, &quot;template to be used after creating a new alignment. &quot;, Validate.REQUIRED);</b>
<i>48</i>&nbsp;	
<i>49</i>&nbsp;	@Override
<i>50</i>&nbsp;	public void initAndValidate() {
<b class="fc"><i>51</i>&nbsp;	}</b>
<i>52</i>&nbsp;	
<i>53</i>&nbsp;	/** 
<i>54</i>&nbsp;	 * return amount to which the provided matches an alignment 
<i>55</i>&nbsp;	 * The provider with the highest match will be used to edit the alignment 
<i>56</i>&nbsp;	 * */
<i>57</i>&nbsp;	protected int matches(Alignment alignment) {
<b class="nc"><i>58</i>&nbsp;		return 1;</b>
<i>59</i>&nbsp;	}
<i>60</i>&nbsp;	
<i>61</i>&nbsp;	/** 
<i>62</i>&nbsp;	 * return new alignment, return null if not successful 
<i>63</i>&nbsp;	 * **/
<i>64</i>&nbsp;	protected List&lt;BEASTInterface&gt; getAlignments(BeautiDoc doc) {
<b class="nc"><i>65</i>&nbsp;        File [] files = beast.app.util.Utils.getLoadFiles(&quot;Load Alignment File&quot;,</b>
<i>66</i>&nbsp;                new File(Beauti.g_sDir), &quot;Alignment files&quot;, &quot;xml&quot;, 
<i>67</i>&nbsp;                &quot;fa&quot;,&quot;fas&quot;,&quot;fst&quot;,&quot;fasta&quot;,&quot;fna&quot;,&quot;ffn&quot;,&quot;faa&quot;,&quot;frn&quot;,
<i>68</i>&nbsp;                &quot;nex&quot;,&quot;nxs&quot;,&quot;nexus&quot;);
<b class="nc"><i>69</i>&nbsp;        if (files != null &amp;&amp; files.length &gt; 0) {</b>
<b class="nc"><i>70</i>&nbsp;            return getAlignments(doc, files);</b>
<i>71</i>&nbsp;        }
<b class="nc"><i>72</i>&nbsp;		return null;</b>
<i>73</i>&nbsp;	}
<i>74</i>&nbsp;
<i>75</i>&nbsp;    /**
<i>76</i>&nbsp;     * return new alignment given files
<i>77</i>&nbsp;     * @param doc
<i>78</i>&nbsp;     * @param files
<i>79</i>&nbsp;     * @return
<i>80</i>&nbsp;     */
<i>81</i>&nbsp;    public List&lt;BEASTInterface&gt; getAlignments(BeautiDoc doc, File[] files) {
<b class="nc"><i>82</i>&nbsp;        List&lt;BEASTInterface&gt; selectedBEASTObjects = new ArrayList&lt;&gt;();</b>
<b class="nc"><i>83</i>&nbsp;        List&lt;MRCAPrior&gt; calibrations = new ArrayList&lt;&gt;();</b>
<b class="nc"><i>84</i>&nbsp;        for (File file : files) {</b>
<b class="nc"><i>85</i>&nbsp;            String fileName = file.getName();</b>
<b class="nc"><i>86</i>&nbsp;			String fileExtension = fileName.substring(fileName.lastIndexOf(&quot;.&quot;)).toLowerCase();</b>
<i>87</i>&nbsp;			Alignment alignment;
<i>88</i>&nbsp;
<b class="nc"><i>89</i>&nbsp;			switch (fileExtension) {</b>
<i>90</i>&nbsp;				case &quot;.nex&quot;:
<i>91</i>&nbsp;				case &quot;.nxs&quot;:
<i>92</i>&nbsp;				case &quot;.nexus&quot;:
<b class="nc"><i>93</i>&nbsp;					NexusParser parser = new NexusParser();</b>
<i>94</i>&nbsp;					try {
<b class="nc"><i>95</i>&nbsp;						parser.parseFile(file);</b>
<b class="nc"><i>96</i>&nbsp;						if (parser.filteredAlignments.size() &gt; 0) {</b>
<i>97</i>&nbsp;							/**
<i>98</i>&nbsp;							 * sanity check: make sure the filters do not
<i>99</i>&nbsp;							 * overlap
<i>100</i>&nbsp;							 **/
<b class="nc"><i>101</i>&nbsp;							int[] used = new int[parser.m_alignment.getSiteCount()];</b>
<b class="nc"><i>102</i>&nbsp;							Set&lt;Integer&gt; overlap = new HashSet&lt;&gt;();</b>
<b class="nc"><i>103</i>&nbsp;							int partitionNr = 1;</b>
<b class="nc"><i>104</i>&nbsp;							for (Alignment data : parser.filteredAlignments) {</b>
<b class="nc"><i>105</i>&nbsp;								int[] indices = ((FilteredAlignment) data).indices();</b>
<b class="nc"><i>106</i>&nbsp;								for (int i : indices) {</b>
<b class="nc"><i>107</i>&nbsp;									if (used[i] &gt; 0) {</b>
<b class="nc"><i>108</i>&nbsp;										overlap.add(used[i] * 10000 + partitionNr);</b>
<i>109</i>&nbsp;									} else {
<b class="nc"><i>110</i>&nbsp;										used[i] = partitionNr;</b>
<i>111</i>&nbsp;									}
<i>112</i>&nbsp;								}
<b class="nc"><i>113</i>&nbsp;								partitionNr++;</b>
<b class="nc"><i>114</i>&nbsp;							}</b>
<b class="nc"><i>115</i>&nbsp;							if (overlap.size() &gt; 0) {</b>
<b class="nc"><i>116</i>&nbsp;								String overlaps = &quot;&lt;html&gt;Warning: The following partitions overlap:&lt;br/&gt;&quot;;</b>
<b class="nc"><i>117</i>&nbsp;								for (int i : overlap) {</b>
<b class="nc"><i>118</i>&nbsp;									overlaps += parser.filteredAlignments.get(i / 10000 - 1).getID()</b>
<i>119</i>&nbsp;											+ &quot; overlaps with &quot;
<b class="nc"><i>120</i>&nbsp;											+ parser.filteredAlignments.get(i % 10000 - 1).getID() + &quot;&lt;br/&gt;&quot;;</b>
<b class="nc"><i>121</i>&nbsp;								}</b>
<b class="nc"><i>122</i>&nbsp;								overlaps += &quot;The first thing you might want to do is delete some of these partitions.&lt;/html&gt;&quot;;</b>
<b class="nc"><i>123</i>&nbsp;								JOptionPane.showMessageDialog(null, overlaps);</b>
<i>124</i>&nbsp;							}
<i>125</i>&nbsp;							/** add alignments **/
<b class="nc"><i>126</i>&nbsp;							for (Alignment data : parser.filteredAlignments) {</b>
<b class="nc"><i>127</i>&nbsp;								sortByTaxonName(data.sequenceInput.get());</b>
<b class="nc"><i>128</i>&nbsp;								selectedBEASTObjects.add(data);</b>
<b class="nc"><i>129</i>&nbsp;							}</b>
<b class="nc"><i>130</i>&nbsp;							if (parser.calibrations != null) {</b>
<b class="nc"><i>131</i>&nbsp;								if (calibrations == null) {</b>
<b class="nc"><i>132</i>&nbsp;									calibrations = new ArrayList&lt;&gt;();</b>
<i>133</i>&nbsp;								}
<b class="nc"><i>134</i>&nbsp;								calibrations.addAll(parser.calibrations);</b>
<i>135</i>&nbsp;							}
<b class="nc"><i>136</i>&nbsp;						} else {</b>
<b class="nc"><i>137</i>&nbsp;							selectedBEASTObjects.add(parser.m_alignment);</b>
<b class="nc"><i>138</i>&nbsp;							if (parser.calibrations != null) {</b>
<b class="nc"><i>139</i>&nbsp;								if (calibrations == null) {</b>
<b class="nc"><i>140</i>&nbsp;									calibrations = new ArrayList&lt;&gt;();</b>
<i>141</i>&nbsp;								}
<b class="nc"><i>142</i>&nbsp;								calibrations.addAll(parser.calibrations);</b>
<i>143</i>&nbsp;							}
<i>144</i>&nbsp;						}
<b class="nc"><i>145</i>&nbsp;					} catch (Exception ex) {</b>
<b class="nc"><i>146</i>&nbsp;						ex.printStackTrace();</b>
<b class="nc"><i>147</i>&nbsp;						JOptionPane.showMessageDialog(null, &quot;Loading of &quot; + fileName + &quot; failed: &quot; + ex.getMessage());</b>
<b class="nc"><i>148</i>&nbsp;						return null;</b>
<b class="nc"><i>149</i>&nbsp;					}</b>
<i>150</i>&nbsp;					break;
<i>151</i>&nbsp;
<i>152</i>&nbsp;				case &quot;.xml&quot;:
<b class="nc"><i>153</i>&nbsp;					alignment = (Alignment)getXMLData(file);</b>
<b class="nc"><i>154</i>&nbsp;					selectedBEASTObjects.add(alignment);</b>
<b class="nc"><i>155</i>&nbsp;					break;</b>
<i>156</i>&nbsp;
<i>157</i>&nbsp;				case &quot;.fa&quot;:
<i>158</i>&nbsp;				case &quot;.fas&quot;:
<i>159</i>&nbsp;				case &quot;.fasta&quot;:
<i>160</i>&nbsp;				case &quot;.fst&quot;:
<i>161</i>&nbsp;				case &quot;.fna&quot;:
<i>162</i>&nbsp;				case &quot;.ffn&quot;:
<i>163</i>&nbsp;				case &quot;.faa&quot;:
<i>164</i>&nbsp;				case &quot;.frn&quot;:
<b class="nc"><i>165</i>&nbsp;					alignment = getFASTAData(file);</b>
<b class="nc"><i>166</i>&nbsp;					sortByTaxonName(alignment.sequenceInput.get());</b>
<b class="nc"><i>167</i>&nbsp;					selectedBEASTObjects.add(alignment);</b>
<b class="nc"><i>168</i>&nbsp;					break;</b>
<i>169</i>&nbsp;
<i>170</i>&nbsp;                default:
<b class="nc"><i>171</i>&nbsp;                    JOptionPane.showMessageDialog(null,</b>
<i>172</i>&nbsp;                            &quot;Unsupported sequence file extension.&quot;,
<i>173</i>&nbsp;                            &quot;Error&quot;, JOptionPane.ERROR_MESSAGE);
<i>174</i>&nbsp;                    break;
<i>175</i>&nbsp;			}
<i>176</i>&nbsp;        }
<b class="nc"><i>177</i>&nbsp;        addAlignments(doc, selectedBEASTObjects);</b>
<b class="nc"><i>178</i>&nbsp;        if (calibrations != null) {</b>
<b class="nc"><i>179</i>&nbsp;        	selectedBEASTObjects.addAll(calibrations);</b>
<i>180</i>&nbsp;        }
<i>181</i>&nbsp;        // doc.addMRCAPriors(calibrations);
<b class="nc"><i>182</i>&nbsp;        return selectedBEASTObjects;</b>
<i>183</i>&nbsp;    }
<i>184</i>&nbsp;    
<i>185</i>&nbsp;    
<i>186</i>&nbsp;    protected void addAlignments(BeautiDoc doc, List&lt;BEASTInterface&gt; selectedBEASTObjects) {
<b class="nc"><i>187</i>&nbsp;        for (BEASTInterface beastObject : selectedBEASTObjects) {</b>
<i>188</i>&nbsp;        	// ensure ID of alignment is unique
<b class="nc"><i>189</i>&nbsp;        	int k = 0;</b>
<b class="nc"><i>190</i>&nbsp;        	String id = beastObject.getID();</b>
<b class="nc"><i>191</i>&nbsp;        	while (doc.pluginmap.containsKey(id)) {</b>
<b class="nc"><i>192</i>&nbsp;        		k++;</b>
<b class="nc"><i>193</i>&nbsp;        		id = beastObject.getID() + k;</b>
<i>194</i>&nbsp;        	}
<b class="nc"><i>195</i>&nbsp;        	beastObject.setID(id);</b>
<b class="nc"><i>196</i>&nbsp;        	sortByTaxonName(((Alignment) beastObject).sequenceInput.get());</b>
<b class="nc"><i>197</i>&nbsp;            doc.addAlignmentWithSubnet((Alignment) beastObject, getStartTemplate());</b>
<b class="nc"><i>198</i>&nbsp;        }</b>
<b class="nc"><i>199</i>&nbsp;    }</b>
<i>200</i>&nbsp;
<i>201</i>&nbsp;	/** provide GUI for manipulating the alignment **/
<i>202</i>&nbsp;	void editAlignment(Alignment alignment, BeautiDoc doc) {
<i>203</i>&nbsp;		try {
<b class="nc"><i>204</i>&nbsp;			AlignmentViewer viewer = new AlignmentViewer(alignment);</b>
<b class="nc"><i>205</i>&nbsp;			viewer.showInDialog();</b>
<b class="nc"><i>206</i>&nbsp;		} catch (Exception e) {</b>
<b class="nc"><i>207</i>&nbsp;			JOptionPane.showMessageDialog(null, &quot;Something went wrong viewing the alignment: &quot; + e.getMessage());</b>
<b class="nc"><i>208</i>&nbsp;			e.printStackTrace();</b>
<b class="nc"><i>209</i>&nbsp;		}</b>
<b class="nc"><i>210</i>&nbsp;	}</b>
<i>211</i>&nbsp;	
<i>212</i>&nbsp;	/** check validity of the alignment, 
<i>213</i>&nbsp;	 * return null if there are no problens, 
<i>214</i>&nbsp;	 * return message string if something is fishy **/
<i>215</i>&nbsp;	String validateAlignment() {
<b class="nc"><i>216</i>&nbsp;		return null;</b>
<i>217</i>&nbsp;	}
<i>218</i>&nbsp;	
<i>219</i>&nbsp;	/** return template to apply to this new alignment.
<i>220</i>&nbsp;	 * By default, the partition template of the current beauti template is returned **/
<i>221</i>&nbsp;	protected BeautiSubTemplate getStartTemplate() {
<b class="nc"><i>222</i>&nbsp;		return template.get();</b>
<i>223</i>&nbsp;	}
<i>224</i>&nbsp;
<i>225</i>&nbsp;    protected void sortByTaxonName(List&lt;Sequence&gt; seqs) {
<b class="nc"><i>226</i>&nbsp;        Collections.sort(seqs, (Sequence o1, Sequence o2) -&gt; {</b>
<b class="nc"><i>227</i>&nbsp;                return o1.taxonInput.get().compareTo(o2.taxonInput.get());</b>
<i>228</i>&nbsp;            }
<i>229</i>&nbsp;        );
<b class="nc"><i>230</i>&nbsp;    }</b>
<i>231</i>&nbsp;
<i>232</i>&nbsp;	static public BEASTInterface getXMLData(File file) {
<b class="nc"><i>233</i>&nbsp;		String xml = &quot;&quot;;</b>
<i>234</i>&nbsp;		try {
<i>235</i>&nbsp;			// parse as BEAST 2 xml fragment
<b class="nc"><i>236</i>&nbsp;			XMLParser parser = new XMLParser();</b>
<b class="nc"><i>237</i>&nbsp;			BufferedReader fin = new BufferedReader(new FileReader(file));</b>
<b class="nc"><i>238</i>&nbsp;			while (fin.ready()) {</b>
<b class="nc"><i>239</i>&nbsp;				xml += fin.readLine() + &quot;\n&quot;;</b>
<i>240</i>&nbsp;			}
<b class="nc"><i>241</i>&nbsp;			fin.close();</b>
<b class="nc"><i>242</i>&nbsp;			BEASTInterface runnable = parser.parseBareFragment(xml, false);</b>
<b class="nc"><i>243</i>&nbsp;			BEASTInterface alignment = getAlignment(runnable);</b>
<b class="nc"><i>244</i>&nbsp;            alignment.initAndValidate();</b>
<b class="nc"><i>245</i>&nbsp;            return alignment;</b>
<b class="nc"><i>246</i>&nbsp;		} catch (Exception ex) {</b>
<i>247</i>&nbsp;			// attempt to parse as BEAST 1 xml
<i>248</i>&nbsp;			try {
<b class="nc"><i>249</i>&nbsp;				String ID = file.getName();</b>
<b class="nc"><i>250</i>&nbsp;				ID = ID.substring(0, ID.lastIndexOf(&#39;.&#39;)).replaceAll(&quot;\\..*&quot;, &quot;&quot;);</b>
<b class="nc"><i>251</i>&nbsp;				BEASTInterface alignment = parseBeast1XML(ID, xml);</b>
<b class="nc"><i>252</i>&nbsp;				if (alignment != null) {</b>
<b class="nc"><i>253</i>&nbsp;					alignment.setID(file.getName().substring(0, file.getName().length() - 4).replaceAll(&quot;\\..*&quot;, &quot;&quot;));</b>
<i>254</i>&nbsp;				}
<b class="nc"><i>255</i>&nbsp;				return alignment;</b>
<b class="nc"><i>256</i>&nbsp;			} catch (Exception ex2) {</b>
<b class="nc"><i>257</i>&nbsp;				ex.printStackTrace();</b>
<b class="nc"><i>258</i>&nbsp;				JOptionPane.showMessageDialog(null, &quot;Loading of &quot; + file.getName() + &quot; failed: &quot; + ex.getMessage()</b>
<b class="nc"><i>259</i>&nbsp;						+ &quot;\n&quot; + ex2.getMessage());</b>
<i>260</i>&nbsp;			}
<b class="nc"><i>261</i>&nbsp;			return null;</b>
<i>262</i>&nbsp;		}
<i>263</i>&nbsp;	}
<i>264</i>&nbsp;	
<i>265</i>&nbsp;    private Alignment getFASTAData(File file) {
<i>266</i>&nbsp;    	try {
<i>267</i>&nbsp;    		// grab alignment data
<b class="nc"><i>268</i>&nbsp;        	Map&lt;String, StringBuilder&gt; seqMap = new HashMap&lt;&gt;();</b>
<b class="nc"><i>269</i>&nbsp;        	List&lt;String&gt; taxa = new ArrayList&lt;&gt;();</b>
<b class="nc"><i>270</i>&nbsp;        	String currentTaxon = null;</b>
<b class="nc"><i>271</i>&nbsp;			BufferedReader fin = new BufferedReader(new FileReader(file));</b>
<b class="nc"><i>272</i>&nbsp;	        String missing = &quot;?&quot;;</b>
<b class="nc"><i>273</i>&nbsp;	        String gap = &quot;-&quot;;</b>
<b class="nc"><i>274</i>&nbsp;	        int totalCount = 4;</b>
<b class="nc"><i>275</i>&nbsp;	        String datatype = &quot;nucleotide&quot;;</b>
<i>276</i>&nbsp;	        // According to http://en.wikipedia.org/wiki/FASTA_format lists file formats and their data content
<i>277</i>&nbsp;			// .fna = nucleic acid
<i>278</i>&nbsp;			// .ffn = nucleotide coding regions
<i>279</i>&nbsp;			// .frn = non-coding RNA
<i>280</i>&nbsp;			// .ffa = amino acid
<b class="nc"><i>281</i>&nbsp;    		boolean mayBeAminoacid = !(file.getName().toLowerCase().endsWith(&quot;.fna&quot;) || file.getName().toLowerCase().endsWith(&quot;.ffn&quot;) || file.getName().toLowerCase().endsWith(&quot;.frn&quot;));</b>
<i>282</i>&nbsp;    		
<b class="nc"><i>283</i>&nbsp;			while (fin.ready()) {</b>
<b class="nc"><i>284</i>&nbsp;				String line = fin.readLine();</b>
<b class="nc"><i>285</i>&nbsp;				if (line.startsWith(&quot;;&quot;)) {</b>
<i>286</i>&nbsp;					// it is a comment, ignore
<b class="nc"><i>287</i>&nbsp;				} else 	if (line.startsWith(&quot;&gt;&quot;)) {</b>
<i>288</i>&nbsp;					// it is a taxon
<b class="nc"><i>289</i>&nbsp;					currentTaxon = line.substring(1).trim();</b>
<i>290</i>&nbsp;					// only up to first space
<b class="nc"><i>291</i>&nbsp;					currentTaxon = currentTaxon.replaceAll(&quot;\\s.*$&quot;, &quot;&quot;);</b>
<i>292</i>&nbsp;				} else {
<i>293</i>&nbsp;					// it is a data line
<b class="nc"><i>294</i>&nbsp;					if (currentTaxon == null) {</b>
<b class="nc"><i>295</i>&nbsp;						fin.close();</b>
<b class="nc"><i>296</i>&nbsp;						throw new RuntimeException(&quot;Expected taxon defined on first line&quot;);</b>
<i>297</i>&nbsp;					}
<b class="nc"><i>298</i>&nbsp;					if (seqMap.containsKey(currentTaxon)) {</b>
<b class="nc"><i>299</i>&nbsp;						StringBuilder sb = seqMap.get(currentTaxon);</b>
<b class="nc"><i>300</i>&nbsp;						sb.append(line);</b>
<b class="nc"><i>301</i>&nbsp;					} else {</b>
<b class="nc"><i>302</i>&nbsp;						StringBuilder sb = new StringBuilder();</b>
<b class="nc"><i>303</i>&nbsp;						seqMap.put(currentTaxon, sb);</b>
<b class="nc"><i>304</i>&nbsp;						sb.append(line);</b>
<b class="nc"><i>305</i>&nbsp;						taxa.add(currentTaxon);</b>
<i>306</i>&nbsp;					}
<i>307</i>&nbsp;				}
<b class="nc"><i>308</i>&nbsp;			}</b>
<b class="nc"><i>309</i>&nbsp;			fin.close();</b>
<i>310</i>&nbsp;			
<b class="nc"><i>311</i>&nbsp;			int charCount = -1;</b>
<b class="nc"><i>312</i>&nbsp;			Alignment alignment = new Alignment();</b>
<b class="nc"><i>313</i>&nbsp;	        for (final String taxon : taxa) {</b>
<b class="nc"><i>314</i>&nbsp;	            final StringBuilder bsData = seqMap.get(taxon);</b>
<b class="nc"><i>315</i>&nbsp;	            String data = bsData.toString();</b>
<b class="nc"><i>316</i>&nbsp;	            data = data.replaceAll(&quot;\\s&quot;, &quot;&quot;);</b>
<b class="nc"><i>317</i>&nbsp;	            seqMap.put(taxon, new StringBuilder(data));</b>
<i>318</i>&nbsp;
<b class="nc"><i>319</i>&nbsp;	            if (charCount &lt; 0) {charCount = data.length();}</b>
<b class="nc"><i>320</i>&nbsp;	            if (data.length() != charCount) {</b>
<b class="nc"><i>321</i>&nbsp;	                throw new IllegalArgumentException(&quot;Expected sequence of length &quot; + charCount + &quot; instead of &quot; + data.length() + &quot; for taxon &quot; + taxon);</b>
<i>322</i>&nbsp;	            }
<i>323</i>&nbsp;	            // map to standard missing and gap chars
<b class="nc"><i>324</i>&nbsp;	            data = data.replace(missing.charAt(0), DataType.MISSING_CHAR);</b>
<b class="nc"><i>325</i>&nbsp;	            data = data.replace(gap.charAt(0), DataType.GAP_CHAR);</b>
<i>326</i>&nbsp;
<b class="nc"><i>327</i>&nbsp;	            if (mayBeAminoacid &amp;&amp; datatype.equals(&quot;nucleotide&quot;) &amp;&amp; !data.matches(&quot;[ACGTUXNacgtuxn?_-]+&quot;)) {</b>
<b class="nc"><i>328</i>&nbsp;	            	datatype = &quot;aminoacid&quot;;</b>
<b class="nc"><i>329</i>&nbsp;	            	totalCount = 20;</b>
<b class="nc"><i>330</i>&nbsp;	            	for (Sequence seq : alignment.sequenceInput.get()) {</b>
<b class="nc"><i>331</i>&nbsp;	            		seq.totalCountInput.setValue(totalCount, seq);</b>
<b class="nc"><i>332</i>&nbsp;	            	}</b>
<i>333</i>&nbsp;	            }
<i>334</i>&nbsp;	            
<b class="nc"><i>335</i>&nbsp;	            final Sequence sequence = new Sequence();</b>
<b class="nc"><i>336</i>&nbsp;	            data = data.replaceAll(&quot;[Xx]&quot;, &quot;?&quot;);</b>
<b class="nc"><i>337</i>&nbsp;	            sequence.init(totalCount, taxon, data);</b>
<b class="nc"><i>338</i>&nbsp;	            sequence.setID(NexusParser.generateSequenceID(taxon));</b>
<b class="nc"><i>339</i>&nbsp;	            alignment.sequenceInput.setValue(sequence, alignment);</b>
<b class="nc"><i>340</i>&nbsp;	        }</b>
<b class="nc"><i>341</i>&nbsp;	        String ID = file.getName();</b>
<b class="nc"><i>342</i>&nbsp;	        ID = ID.substring(0, ID.lastIndexOf(&#39;.&#39;)).replaceAll(&quot;\\..*&quot;, &quot;&quot;);</b>
<b class="nc"><i>343</i>&nbsp;	        alignment.setID(ID);</b>
<b class="nc"><i>344</i>&nbsp;			alignment.dataTypeInput.setValue(datatype, alignment);</b>
<b class="nc"><i>345</i>&nbsp;	        alignment.initAndValidate();</b>
<b class="nc"><i>346</i>&nbsp;	        return alignment;</b>
<b class="nc"><i>347</i>&nbsp;    	} catch (Exception e) {</b>
<b class="nc"><i>348</i>&nbsp;			e.printStackTrace();</b>
<b class="nc"><i>349</i>&nbsp;			JOptionPane.showMessageDialog(null, &quot;Loading of &quot; + file.getName() + &quot; failed: &quot; + e.getMessage());</b>
<i>350</i>&nbsp;    	}
<b class="nc"><i>351</i>&nbsp;		return null;</b>
<i>352</i>&nbsp;	}
<i>353</i>&nbsp;
<i>354</i>&nbsp;	private static BEASTInterface parseBeast1XML(String ID, String xml) throws SAXException, IOException, ParserConfigurationException  {
<b class="nc"><i>355</i>&nbsp;		DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();</b>
<b class="nc"><i>356</i>&nbsp;		Document doc = factory.newDocumentBuilder().parse(new InputSource(new StringReader(xml)));</b>
<b class="nc"><i>357</i>&nbsp;		doc.normalize();</b>
<i>358</i>&nbsp;
<b class="nc"><i>359</i>&nbsp;		NodeList alignments = doc.getElementsByTagName(&quot;alignment&quot;);</b>
<b class="nc"><i>360</i>&nbsp;		Alignment alignment = new Alignment();</b>
<b class="nc"><i>361</i>&nbsp;		alignment.dataTypeInput.setValue(&quot;nucleotide&quot;, alignment);</b>
<i>362</i>&nbsp;
<i>363</i>&nbsp;		// parse first alignment
<b class="nc"><i>364</i>&nbsp;		org.w3c.dom.Node node = alignments.item(0);</b>
<i>365</i>&nbsp;
<b class="nc"><i>366</i>&nbsp;		String dataTypeName = node.getAttributes().getNamedItem(&quot;dataType&quot;).getNodeValue();</b>
<b class="nc"><i>367</i>&nbsp;		int totalCount = 4;</b>
<b class="nc"><i>368</i>&nbsp;		if (dataTypeName == null) {</b>
<b class="nc"><i>369</i>&nbsp;			alignment.dataTypeInput.setValue(&quot;integer&quot;, alignment);</b>
<b class="nc"><i>370</i>&nbsp;		} else if (dataTypeName.toLowerCase().equals(&quot;dna&quot;) || dataTypeName.toLowerCase().equals(&quot;nucleotide&quot;)) {</b>
<b class="nc"><i>371</i>&nbsp;			alignment.dataTypeInput.setValue(&quot;nucleotide&quot;, alignment);</b>
<b class="nc"><i>372</i>&nbsp;			totalCount = 4;</b>
<b class="nc"><i>373</i>&nbsp;		} else if (dataTypeName.toLowerCase().equals(&quot;aminoacid&quot;) || dataTypeName.toLowerCase().equals(&quot;protein&quot;)) {</b>
<b class="nc"><i>374</i>&nbsp;			alignment.dataTypeInput.setValue(&quot;aminoacid&quot;, alignment);</b>
<b class="nc"><i>375</i>&nbsp;			totalCount = 20;</b>
<i>376</i>&nbsp;		} else {
<b class="nc"><i>377</i>&nbsp;			alignment.dataTypeInput.setValue(&quot;integer&quot;, alignment);</b>
<i>378</i>&nbsp;		}
<i>379</i>&nbsp;
<b class="nc"><i>380</i>&nbsp;		NodeList children = node.getChildNodes();</b>
<b class="nc"><i>381</i>&nbsp;		for (int i = 0; i &lt; children.getLength(); i++) {</b>
<b class="nc"><i>382</i>&nbsp;			org.w3c.dom.Node child = children.item(i);</b>
<b class="nc"><i>383</i>&nbsp;			if (child.getNodeName().equals(&quot;sequence&quot;)) {</b>
<b class="nc"><i>384</i>&nbsp;				Sequence sequence = new Sequence();</b>
<i>385</i>&nbsp;				// find the taxon
<b class="nc"><i>386</i>&nbsp;				String taxon = &quot;&quot;;</b>
<b class="nc"><i>387</i>&nbsp;				NodeList sequenceChildren = child.getChildNodes();</b>
<b class="nc"><i>388</i>&nbsp;				for (int j = 0; j &lt; sequenceChildren.getLength(); j++) {</b>
<b class="nc"><i>389</i>&nbsp;					org.w3c.dom.Node child2 = sequenceChildren.item(j);</b>
<b class="nc"><i>390</i>&nbsp;					if (child2.getNodeName().equals(&quot;taxon&quot;)) {</b>
<b class="nc"><i>391</i>&nbsp;						taxon = child2.getAttributes().getNamedItem(&quot;idref&quot;).getNodeValue();</b>
<i>392</i>&nbsp;					}
<i>393</i>&nbsp;				}
<b class="nc"><i>394</i>&nbsp;				String data = child.getTextContent();</b>
<b class="nc"><i>395</i>&nbsp;				sequence.initByName(&quot;totalcount&quot;, totalCount, &quot;taxon&quot;, taxon, &quot;value&quot;, data);</b>
<b class="nc"><i>396</i>&nbsp;				sequence.setID(&quot;seq_&quot; + taxon);</b>
<b class="nc"><i>397</i>&nbsp;				alignment.sequenceInput.setValue(sequence, alignment);</b>
<i>398</i>&nbsp;
<i>399</i>&nbsp;			}
<i>400</i>&nbsp;		}
<b class="nc"><i>401</i>&nbsp;		alignment.setID(ID);</b>
<b class="nc"><i>402</i>&nbsp;		alignment.initAndValidate();</b>
<b class="nc"><i>403</i>&nbsp;		return alignment;</b>
<i>404</i>&nbsp;	} // parseBeast1XML
<i>405</i>&nbsp;
<i>406</i>&nbsp;
<i>407</i>&nbsp;	static BEASTInterface getAlignment(BEASTInterface beastObject) throws IllegalArgumentException, IllegalAccessException {
<b class="nc"><i>408</i>&nbsp;		if (beastObject instanceof Alignment) {</b>
<b class="nc"><i>409</i>&nbsp;			return beastObject;</b>
<i>410</i>&nbsp;		}
<b class="nc"><i>411</i>&nbsp;		for (BEASTInterface beastObject2 : beastObject.listActiveBEASTObjects()) {</b>
<b class="nc"><i>412</i>&nbsp;			beastObject2 = getAlignment(beastObject2);</b>
<b class="nc"><i>413</i>&nbsp;			if (beastObject2 != null) {</b>
<b class="nc"><i>414</i>&nbsp;				return beastObject2;</b>
<i>415</i>&nbsp;			}
<b class="nc"><i>416</i>&nbsp;		}</b>
<b class="nc"><i>417</i>&nbsp;		return null;</b>
<i>418</i>&nbsp;	}
<i>419</i>&nbsp;
<i>420</i>&nbsp;}
</div>
</div>

<div class="footer">
    
    <div style="float:right;">generated on 2016-05-30 16:46</div>
</div>
</body>
</html>
